common/phal_F4/gpio/gpio.h

```diff
@@ -0,0 +1,228 @@
+    {                                                                                                \
+        .bank = gpio_bank, .pin = pin_num, .type = GPIO_TYPE_AF, .config = {.af_num = alt_func_num,  \
+                                                                            .ospeed = ospeed_sel,    \
+                                                                            .otype = otype_sel,      \
+                                                                            .pull = input_pull_sel } \
+    }
+
+/*
+    Useful defines for GPIO Init struct with commonly used peripheral/pin mappings.
+    If you find yourself adding the same pin mappings to multiple devices, add a macro below
+    to cut down on duplication.
+*/
+#define GPIO_INIT_CANRX_PA11 GPIO_INIT_AF(GPIOA, 11, 9, GPIO_OUTPUT_ULTRA_SPEED, GPIO_TYPE_AF, GPIO_INPUT_OPEN_DRAIN)
+#define GPIO_INIT_CANTX_PA12 GPIO_INIT_AF(GPIOA, 12, 9, GPIO_OUTPUT_ULTRA_SPEED, GPIO_TYPE_AF, GPIO_INPUT_OPEN_DRAIN)
+#define GPIO_INIT_CANRX_PD0 GPIO_INIT_AF(GPIOD, 0, 9, GPIO_OUTPUT_ULTRA_SPEED, GPIO_TYPE_AF, GPIO_INPUT_OPEN_DRAIN)
+#define GPIO_INIT_CANTX_PD1 GPIO_INIT_AF(GPIOD, 1, 9, GPIO_OUTPUT_ULTRA_SPEED, GPIO_TYPE_AF, GPIO_INPUT_OPEN_DRAIN)
+#define GPIO_INIT_CAN2RX_PB12 GPIO_INIT_AF(GPIOB, 12, 10, GPIO_OUTPUT_ULTRA_SPEED, GPIO_TYPE_AF, GPIO_INPUT_OPEN_DRAIN)
+#define GPIO_INIT_CAN2TX_PB13 GPIO_INIT_AF(GPIOB, 13, 10, GPIO_OUTPUT_ULTRA_SPEED, GPIO_TYPE_AF, GPIO_INPUT_OPEN_DRAIN)
+#define GPIO_INIT_I2C1_SCL_PA9 GPIO_INIT_AF(GPIOA, 9, 4, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_OPEN_DRAIN, GPIO_INPUT_PULL_UP)
+#define GPIO_INIT_I2C1_SDA_PA10 GPIO_INIT_AF(GPIOA, 10, 4, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_OPEN_DRAIN, GPIO_INPUT_PULL_UP)
+#define GPIO_INIT_I2C1_SCL_PB6 GPIO_INIT_AF(GPIOB, 6, 4, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_OPEN_DRAIN, GPIO_INPUT_PULL_UP)
+#define GPIO_INIT_I2C1_SDA_PB7 GPIO_INIT_AF(GPIOB, 7, 4, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_OPEN_DRAIN, GPIO_INPUT_PULL_UP)
+#define GPIO_INIT_I2C3_SCL_PA7 GPIO_INIT_AF(GPIOA, 7, 4, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_OPEN_DRAIN, GPIO_INPUT_PULL_UP)
+#define GPIO_INIT_I2C3_SDA_PB4 GPIO_INIT_AF(GPIOB, 4, 4, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_OPEN_DRAIN, GPIO_INPUT_PULL_UP)
+#define GPIO_INIT_I2C4_SCL_PB10 GPIO_INIT_AF(GPIOB, 10, 4, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_OPEN_DRAIN, GPIO_INPUT_PULL_UP)
+#define GPIO_INIT_I2C4_SDA_PB11 GPIO_INIT_AF(GPIOB, 11, 4, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_OPEN_DRAIN, GPIO_INPUT_PULL_UP)
+#define GPIO_INIT_USART1TX_PA9 GPIO_INIT_AF(GPIOA, 9, 7, GPIO_OUTPUT_ULTRA_SPEED, GPIO_TYPE_AF, GPIO_INPUT_OPEN_DRAIN)
+#define GPIO_INIT_USART1RX_PA10 GPIO_INIT_AF(GPIOA, 10, 7, GPIO_OUTPUT_ULTRA_SPEED, GPIO_TYPE_AF, GPIO_INPUT_OPEN_DRAIN)
+#define GPIO_INIT_USART1TX_PB6 GPIO_INIT_AF(GPIOB, 6, 7, GPIO_OUTPUT_ULTRA_SPEED, GPIO_TYPE_AF, GPIO_INPUT_OPEN_DRAIN)
+#define GPIO_INIT_USART1RX_PB7 GPIO_INIT_AF(GPIOB, 7, 7, GPIO_OUTPUT_ULTRA_SPEED, GPIO_TYPE_AF, GPIO_INPUT_OPEN_DRAIN)
```

Comment:
Are all these pin defs valid? Better to have no definition than wrong definition.